### ISSUE TITLE
Label data corrupted after import -D

### ISSUE THREAD

**Issue by lynchc** [zfsonlinux/zfs#3024]
My usb3 controller died over the weekend and dropped connection to my enclosure where my zpool lives. After it rebooted, the pool didn't remount, which isn't that weird. I ran a zpool import tank and it mounted the pool but none of the folders within the pool had any data in them. I could only see the root-level folders and then nothing inside. I exported and tried to import again and that didn't work.
Then I ran a zpool import -D tank (by accident mostly because I forgot the option) and it imported an older tank that used three of the drives from the tank I wanted.

So now, I'm in a situation where I can see two zpools, one destroyed that can access those three drives, and my main pool that has all of the drives but cannot access those three drives. I looked at the label data on those three drives and it's been over-written to the old pool. How do I fix this? is there a way to re-import these three drives back into their proper pool somehow? is there a command I can issue in zdb that will rewrite this metadata?

**Comment by dasjoe** [comment]
Please paste a "zpool import" with all pools exported

**Comment by lynchc** [comment]
zpool import
pool: storage
     id: 15495756392794434226
  state: UNAVAIL
 status: One or more devices contains corrupted data.
 action: The pool cannot be imported due to damaged devices or data.
   see: http://zfsonlinux.org/msg/ZFS-8000-5E
 config:

	storage                              UNAVAIL  insufficient replicas
	  raidz1-0                           UNAVAIL  insufficient replicas
	    ata-ST3000DM001-1CH166_Z1F1PYM6  UNAVAIL
	    ata-ST3000DM001-1ER166_Z5003V9R  ONLINE
	    ata-ST3000DM001-1CH166_W1F2372R  UNAVAIL
	    ata-ST3000DM001-1CH166_W1F24BTK  UNAVAIL
	    ata-ST3000DM001-1CH166_Z1F2KKLW  ONLINE
	  mirror-1                           ONLINE
	    ata-ST3000DM001-1ER166_Z5003Q9H  ONLINE
	    ata-ST3000DM001-1ER166_Z7P0077P  ONLINE

# zpool import -D
   pool: storage
     id: 7750158761011452865
  state: FAULTED (DESTROYED)
 status: One or more devices contains corrupted data.
 action: The pool cannot be imported due to damaged devices or data.
	The pool was destroyed, but can be imported using the '-Df' flags.
   see: http://zfsonlinux.org/msg/ZFS-8000-5E
 config:

	storage                              FAULTED  corrupted data
	  raidz1-0                           DEGRADED
	    ata-ST3000DM001-1CH166_Z1F1PYM6  UNAVAIL  corrupted data
	    10489509276383395998             UNAVAIL
	    ata-ST3000DM001-1CH166_W1F2372R  ONLINE
	    ata-ST3000DM001-1CH166_W1F24BTK  ONLINE
root@mainserver:/# zdb -l /dev/disk/by-id/ata-ST3000DM001-1CH166_W1F24BTK
--------------------------------------------
LABEL 0
--------------------------------------------
    version: 28
    name: 'storage'
    state: 2
    txg: 205
    pool_guid: 7750158761011452865
    hostname: 'mainserver'
    top_guid: 13046049972944463509
    guid: 1576658373984103301
    vdev_children: 1
    vdev_tree:
        type: 'raidz'
        id: 0
        guid: 13046049972944463509
        nparity: 1
        metaslab_array: 30
        metaslab_shift: 36
        ashift: 9
        asize: 12002352168960
        is_log: 0
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 2643899950892783677
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F1PYM6'
            phys_path: '/dev/da0'
            whole_disk: 1
            create_txg: 4
        children[1]:
            type: 'disk'
            id: 1
            guid: 10489509276383395998
            path: '/dev/da1'
            phys_path: '/dev/da1'
            whole_disk: 1
            not_present: 1
            DTL: 41
            create_txg: 4
        children[2]:
            type: 'disk'
            id: 2
            guid: 12725331665777688069
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F2372R'
            phys_path: '/dev/da2'
            whole_disk: 1
            create_txg: 4
        children[3]:
            type: 'disk'
            id: 3
            guid: 1576658373984103301
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F24BTK'
            phys_path: '/dev/da3'
            whole_disk: 1
            create_txg: 4
    features_for_read:
--------------------------------------------
LABEL 1
--------------------------------------------
    version: 28
    name: 'storage'
    state: 2
    txg: 205
    pool_guid: 7750158761011452865
    hostname: 'mainserver'
    top_guid: 13046049972944463509
    guid: 1576658373984103301
    vdev_children: 1
    vdev_tree:
        type: 'raidz'
        id: 0
        guid: 13046049972944463509
        nparity: 1
        metaslab_array: 30
        metaslab_shift: 36
        ashift: 9
        asize: 12002352168960
        is_log: 0
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 2643899950892783677
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F1PYM6'
            phys_path: '/dev/da0'
            whole_disk: 1
            create_txg: 4
        children[1]:
            type: 'disk'
            id: 1
            guid: 10489509276383395998
            path: '/dev/da1'
            phys_path: '/dev/da1'
            whole_disk: 1
            not_present: 1
            DTL: 41
            create_txg: 4
        children[2]:
            type: 'disk'
            id: 2
            guid: 12725331665777688069
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F2372R'
            phys_path: '/dev/da2'
            whole_disk: 1
            create_txg: 4
        children[3]:
            type: 'disk'
            id: 3
            guid: 1576658373984103301
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F24BTK'
            phys_path: '/dev/da3'
            whole_disk: 1
            create_txg: 4
    features_for_read:
--------------------------------------------
LABEL 2
--------------------------------------------
    version: 28
    name: 'storage'
    state: 2
    txg: 205
    pool_guid: 7750158761011452865
    hostname: 'mainserver'
    top_guid: 13046049972944463509
    guid: 1576658373984103301
    vdev_children: 1
    vdev_tree:
        type: 'raidz'
        id: 0
        guid: 13046049972944463509
        nparity: 1
        metaslab_array: 30
        metaslab_shift: 36
        ashift: 9
        asize: 12002352168960
        is_log: 0
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 2643899950892783677
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F1PYM6'
            phys_path: '/dev/da0'
            whole_disk: 1
            create_txg: 4
        children[1]:
            type: 'disk'
            id: 1
            guid: 10489509276383395998
            path: '/dev/da1'
            phys_path: '/dev/da1'
            whole_disk: 1
            not_present: 1
            DTL: 41
            create_txg: 4
        children[2]:
            type: 'disk'
            id: 2
            guid: 12725331665777688069
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F2372R'
            phys_path: '/dev/da2'
            whole_disk: 1
            create_txg: 4
        children[3]:
            type: 'disk'
            id: 3
            guid: 1576658373984103301
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F24BTK'
            phys_path: '/dev/da3'
            whole_disk: 1
            create_txg: 4
    features_for_read:
--------------------------------------------
LABEL 3
--------------------------------------------
    version: 28
    name: 'storage'
    state: 2
    txg: 205
    pool_guid: 7750158761011452865
    hostname: 'mainserver'
    top_guid: 13046049972944463509
    guid: 1576658373984103301
    vdev_children: 1
    vdev_tree:
        type: 'raidz'
        id: 0
        guid: 13046049972944463509
        nparity: 1
        metaslab_array: 30
        metaslab_shift: 36
        ashift: 9
        asize: 12002352168960
        is_log: 0
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 2643899950892783677
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F1PYM6'
            phys_path: '/dev/da0'
            whole_disk: 1
            create_txg: 4
        children[1]:
            type: 'disk'
            id: 1
            guid: 10489509276383395998
            path: '/dev/da1'
            phys_path: '/dev/da1'
            whole_disk: 1
            not_present: 1
            DTL: 41
            create_txg: 4
        children[2]:
            type: 'disk'
            id: 2
            guid: 12725331665777688069
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F2372R'
            phys_path: '/dev/da2'
            whole_disk: 1
            create_txg: 4
        children[3]:
            type: 'disk'
            id: 3
            guid: 1576658373984103301
            path: '/dev/disk/by-id/ata-ST3000DM001-1CH166_W1F24BTK'
            phys_path: '/dev/da3'
            whole_disk: 1
            create_txg: 4
    features_for_read:

# zdb -l /dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F2KKLW
--------------------------------------------
LABEL 0
--------------------------------------------
failed to unpack label 0
--------------------------------------------
LABEL 1
--------------------------------------------
failed to unpack label 1
--------------------------------------------
LABEL 2
--------------------------------------------
    version: 28
    name: 'storage'
    state: 1
    txg: 3706162
    pool_guid: 15495756392794434226
    hostid: 986977394
    hostname: 'mainserver'
    top_guid: 12384024430068692554
    guid: 13128615732339312
    vdev_children: 1
    vdev_tree:
        type: 'raidz'
        id: 0
        guid: 12384024430068692554
        nparity: 1
        metaslab_array: 30
        metaslab_shift: 37
        ashift: 12
        asize: 15002934968320
        is_log: 0
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 2142571504329829734
            path: '/dev/da0p1'
            phys_path: '/dev/da0p1'
            whole_disk: 1
            DTL: 72
            create_txg: 4
        children[1]:
            type: 'disk'
            id: 1
            guid: 9444550020518283754
            path: '/dev/gpt/disk01'
            phys_path: '/dev/gpt/disk01'
            whole_disk: 1
            DTL: 71
            create_txg: 4
        children[2]:
            type: 'disk'
            id: 2
            guid: 6314456135545678959
            path: '/dev/da2p1'
            phys_path: '/dev/da2p1'
            whole_disk: 1
            DTL: 70
            create_txg: 4
        children[3]:
            type: 'disk'
            id: 3
            guid: 2399075792257730622
            path: '/dev/da3p1'
            phys_path: '/dev/da3p1'
            whole_disk: 1
            DTL: 69
            create_txg: 4
        children[4]:
            type: 'disk'
            id: 4
            guid: 13128615732339312
            path: '/dev/gptid/af72ff8a-be49-11e2-8f4b-001aa0a7a8b2'
            phys_path: '/dev/gptid/af72ff8a-be49-11e2-8f4b-001aa0a7a8b2'
            whole_disk: 1
            DTL: 68
            create_txg: 4
--------------------------------------------
LABEL 3
--------------------------------------------
    version: 28
    name: 'storage'
    state: 1
    txg: 3706162
    pool_guid: 15495756392794434226
    hostid: 986977394
    hostname: 'mainserver'
    top_guid: 12384024430068692554
    guid: 13128615732339312
    vdev_children: 1
    vdev_tree:
        type: 'raidz'
        id: 0
        guid: 12384024430068692554
        nparity: 1
        metaslab_array: 30
        metaslab_shift: 37
        ashift: 12
        asize: 15002934968320
        is_log: 0
        create_txg: 4
        children[0]:
            type: 'disk'
            id: 0
            guid: 2142571504329829734
            path: '/dev/da0p1'
            phys_path: '/dev/da0p1'
            whole_disk: 1
            DTL: 72
            create_txg: 4
        children[1]:
            type: 'disk'
            id: 1
            guid: 9444550020518283754
            path: '/dev/gpt/disk01'
            phys_path: '/dev/gpt/disk01'
            whole_disk: 1
            DTL: 71
            create_txg: 4
        children[2]:
            type: 'disk'
            id: 2
            guid: 6314456135545678959
            path: '/dev/da2p1'
            phys_path: '/dev/da2p1'
            whole_disk: 1
            DTL: 70
            create_txg: 4
        children[3]:
            type: 'disk'
            id: 3
            guid: 2399075792257730622
            path: '/dev/da3p1'
            phys_path: '/dev/da3p1'
            whole_disk: 1
            DTL: 69
            create_txg: 4
        children[4]:
            type: 'disk'
            id: 4
            guid: 13128615732339312
            path: '/dev/gptid/af72ff8a-be49-11e2-8f4b-001aa0a7a8b2'
            phys_path: '/dev/gptid/af72ff8a-be49-11e2-8f4b-001aa0a7a8b2'
            whole_disk: 1
            DTL: 68
            create_txg: 4

**Comment by dasjoe** [comment]
So, you have two pools, both using whole disks.

One is ushing ashift=9, the other ashift=12.

Somehow importing the previously destroyed pool (which was using ashift=9) overwrote the old pool's disks' labels without warning(?), I doubt this should happen.

Any idea why labels 1 and 2 are missing for `/dev/disk/by-id/ata-ST3000DM001-1CH166_Z1F2KKLW`?

What does `zdb -l /dev/disk/by-id/ata-ST3000DM001-1CH166_W1F24BTK-part1` give? If that's showing the new pool's vdev you might have success with something like:

```
sudo -i
mkdir tmp-pool
cd tmp-pool
for disk in /dev/sd*1; do ln -s "$disk" .; done
zpool import -d $(pwd) 15495756392794434226
```

**Comment by lynchc** [comment]
There is no 24btk-part1

**Comment by lynchc** [comment]
solved. ryao from zfsonlinux tracked down the issue for me

**Comment by dasjoe** [comment]
Could you give a detailed summary of what happened and how you fixed your disks' labels?

**Comment by ryao** [comment]
@dasjoe The disks had originally been used on a partition-free pool made on FreeBSD that had been destroyed. That pool had the same name "storage". When `zpool import -D storage` was run, it attempted to import this pool, which failed with the pool marked faulted. Exporting that pool did a label update, which clobbered the GPT partition tables. I walked @lynchc through the process of repairing the partition tables. I made the assumption that the all disks had identical partition tables and told @lynchc to use sgdisk to copy the partition tables from the good disks to the bad ones. This allowed the actual pool to be imported. It turned out that the "good" disks had partially clobbered GPT tables too because of a GPT checksum failure, but the rear labels were okay such that gdisk could repair the tables and all was fine.
